text/1004-transactions-api.rst

```diff
@@ -403,107 +601,197 @@ where we catch network error before sending a request.
+   specified timeout each time.
+
+Only the following conditions are treated as eligible for reconnect:
+
+1. Name resolution failed
+2. File not found (Unix socket not bound yet)
+3. Connection reset, connection aborted, connection refused (server is
+   restarting or not ready yet)
+4. Timeout happened during connect or authentication
+
+All other errors are propagated immediately.
+
+This timeout is different from ``connect_timeout``. Connect timeout
+determines how long individual connect attempt may take. And
+``wait_until_available`` determines how many such attempts could be
+made (i.e. how many ones fits the time frame).
+
+
+EdgeDB Changes
+--------------
+
+To support features above we add two headers to EdgeDB queries:
 
-While this covers "concurent update" and "deadlock" errors, with the
-current implementation it can't reconnect when when connection was
-interrupted while transaction is active. There are few ways it can work:
+1. For PrepareComplete_, CommandComplete_ server-side messages:
+   ``SERVER_OPT_HAS_FEATURES``
+2. For Prepare_, OptimisticExecute_, Execute_, ExecuteScript_
+   client-side messages: ``QUERY_OPT_ALLOW_FEATURES``
 
-a. It may only retry on the same connection and fail on disconnect
-b. It may reconnect and replace underlying socket in the connection
-   object and retry
-c. We may only allow transactions on connection pools (current
-   specification)
+Both contain the set of strings joined by comma:
 
-This can be postponed to later RFC when we know what are use cases of
-it.
+1. `modification` -- query is not read-only
+2. `session` -- query contains session config change
+3. `config` -- server or database config change
+4. `ddl` -- query contains DDL
+5. `transaction` -- query contains start/commit/rollback of transaction
+    or savepoint manipulation
 
+In case of ``SERVER_OPT_HAS_FEATURES`` it describes what is actually
+contained in the query. And in case of ``QUERY_OPT_ALLOW_FEATURES``
+client can specify what of these things are allowed in this query.
 
-Retry Single Queries
---------------------
+Read-only queries are always allowed. When ``QUERY_OPT_ALLOW_FEATURES``
+is omitted any query is allowed (default). With the empty string only
+read-only queries are allowed.
 
-While it's tempting to retry ``pool.query`` and ``pool.execute`` calls,
-it **gives the false sense of security**: no "concurrent update" issues
-seen.  But it's better to see such error and turn the whole block of
-code into a transaction rather than just a mutation. I.e. retrying a
-single mutable request on a "concurrent update" error must be a
-deliberate decision.
+``SERVER_OPT_HAS_FEATURES`` is an empty string for read-only queries (the
+field is present) as it indicates that query has been analyzed.
 
+The ``SERVER_OPT_HAS_FEATURES`` is needed for the following tasks:
 
-Retry Single Queries on Connection Error
-----------------------------------------
+1. Retry standalone (non-transactional) read-only queries
```

Comment:
You mean that the server would retry some of the client's queries on its own? I wouldn't like that.

Reply:
Not server, client.